### PR TITLE
 feat: Enhanced Different TTLSecondsAfterFinished depending on if job is in Succeeded, Failed or Error, Fixes

### LINKED ISSUE #1827: Different TTLSecondsAfterFinished depending on if job is in Succeeded, Failed or Error?
**Is this a BUG REPORT or FEATURE REQUEST?**:
FEATURE REQUEST

**What happened**:

Argo workflows [TTLSecondsAfterFinished](https://github.com/argoproj/argo/blob/5274afb97686a4d2a58c50c3b23dd2b680b881e6/pkg/apis/workflow/v1alpha1/workflow_types.go#L202) offers the possibility so handle clean up of workflows that have finished execution. Which means workflows in either Succeeded, Failed or Error state. My question  about the background to setting ```Succeeded, Failed, Error```to have the same time to live.  

I am currently setting up [Kubeflow](https://www.kubeflow.org) which realise on Argo for what is called [Kubeflow Pipelines](https://github.com/kubeflow/pipelines). In order to make sure we don't end up with a lot of old finished/failed workflows we use the TTLSecondsAfterFinished (have also set up a mutating admission web hook to always default and check if users are within the allowed time window, happy to contribute this but will make a separate issue. ) However we have been in discussions in my team about if we should have different time to live depending on if the workflow fails or succeed. The main discussions point is that we would like to extend the time to live for workflows in the failed state to allow for debugging during a longer time window.  How was the thought when this was set up? Would it be of interest to add a separation between Succeeded, Failed and Error? 

**What you expected to happen**:
 FEATURE REQUEST  -- nothing to add here

**How to reproduce it (as minimally and precisely as possible)**:
 FEATURE REQUEST  -- nothing to add here

**Anything else we need to know?**:
I am happy to make contributions depending on your feedback and thoughts. 

**Environment**:
- Argo version:
```
$ argo version
```
- Kubernetes version :
```
$ kubectl version -o yaml
```

**Other debugging information (if applicable)**:
- workflow result:
```
$ argo get <workflowname>
```
- executor logs:
```
$ kubectl logs <failedpodname> -c init
$ kubectl logs <failedpodname> -c wait
```
- workflow-controller logs:
```
$ kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

### PR DESCRIPTION
**Pull request by NikeNano**
Checklist:

I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/1827) and discussed it with the community. 

Comment: This PR is note done but I have added it in order to get some more specific feedback and discuss the implementation.

### CODE REVIEW

**Comment by NikeNano** [comment]
Hi @sarabala1979 I have now made an initial PR with the workflow spec setup. I would be great if you could give some initial feedback. 

Thoughts: 
- I decided to leave TTLSecondsAfterFinished and added TTLStragety as : 
```
type TTLStrategy struct {
	SecondsAfterSuccess *int32 `json:"SecondsAfterSuccess:,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterSuccess:"`
	SecondsAfterFailed  *int32 `json:"SecondsAfterFailed,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterFailed"`
} 
```
My thought was that now there is now no change to the current design of TTLSecondsAfterFinished which should improve backward compatibility. 
- I also added so that priority is given for SecondsAfterSuccess and SecondsAfterFailed.  I updated with tests as well. 

Concerning the configmap for the controller. I have started to look for it, is it correct that what you mean by configmap for the controller is (this)[https://github.com/argoproj/argo/blob/ce78227abe5a3c901e5b7a7dd823fb2551dff584/workflow/config/config.go#L10]. 

If so how do you think is the base way to pass the information from the configmap object to the ttlcontroller? Atm as I understand the configmap is not passed to the ttlcontroller. 

Thanks for the feedback!

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
 var _ TemplateGetter = &Workflow{}
 var _ TemplateStorage = &Workflow{}
 
+// TTLStrategy is the strategy for the time to live depending on if the workflow succeded or failed
+type TTLStrategy struct {
+	SecondsAfterSuccess *int32 `json:"SecondsAfterSuccess:,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterSuccess:"`
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: Our convention has `json` variables be `camelCase`. Also, start with a fresh `protobuf` index for each `struct`
```suggestion
	SecondsAfterSuccess *int32 `json:"secondsAfterSuccess:,omitempty" protobuf:"bytes,1,opt,name=SecondsAfterSuccess:"`
```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
...
+	SecondsAfterFailed  *int32 `json:"SecondsAfterFailed,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterFailed"`
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: Likewise
```suggestion
	SecondsAfterFailed  *int32 `json:"secondsAfterFailed,omitempty" protobuf:"bytes,2,opt,name=SecondsAfterFailed"`
```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -201,6 +208,13 @@ type WorkflowSpec struct {
 	// ttlSecondsAfterFinished expires immediately after the Workflow finishes.
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"bytes,18,opt,name=ttlSecondsAfterFinished"`
 
+	// TTLStrategy limits the lifetime of a Workflow that has finished execution depending on if it
+	// Succeeded or Failed. If this struct is set, once the Workflow finishes, it will be
+	// deleted after the time to live expires. If this field is unset,
+	// the controller config map will hold the default values
+	// Update
+	TTLStrategy TTLStrategy `json:"TTLStrategy,omitempty" protobuf:"bytes,18,opt,name=TTLStrategy"`
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]:
```suggestion
	TTLStrategy TTLStrategy `json:"ttlStrategy,omitempty" protobuf:"bytes,28,opt,name=ttlStrategy"`
```

Comment (simster7) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: This field should also be of type `*TTLStrategy`

File: workflow/ttlcontroller/ttlcontroller.go [hunk 4]
```diff
@@ -129,6 +129,9 @@ func (c *Controller) processNextWorkItem() bool {
 }
 
 // enqueueWF conditionally queues a workflow to the ttl queue if it is within the deletion period
+// what is deletion period?
+// need to update here as well it seams like .....
+
```
Comment (simster7) [inline workflow/ttlcontroller/ttlcontroller.go]: Did you mean to delete these?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 5]
```diff
@@ -855,6 +869,11 @@ func (ws *WorkflowStatus) Successful() bool {
 	return ws.Phase == NodeSucceeded
 }
 
+// Failed return where or not the workflow has failed
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]:
```suggestion
// Failed return whether or not the workflow has failed
```

File: workflow/ttlcontroller/ttlcontroller.go [hunk 6]
```diff
@@ -199,24 +202,43 @@ func (c *Controller) deleteWorkflow(key string) error {
...
+	} else if wf.Status.Successful() && wf.Spec.TTLStrategy.SecondsAfterSuccess != nil {
+		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*wf.Spec.TTLStrategy.SecondsAfterSuccess))
+		return now.After(expiry)
+	} else {
+		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*wf.Spec.TTLSecondsAfterFinished))
+		return now.After(expiry)
+	}
 }
 
+// This is used to put thing on the que
```
Comment (simster7) [inline workflow/ttlcontroller/ttlcontroller.go]: Did you mean to delete this one too?

**Comment by NikeNano** [comment]
Thanks for the feedback @simster7  I Will update based upon  your suggestions :)

**Comment by NikeNano** [comment]
> Thanks for your PR @NikeNano! A couple of changes needed.
> 
> Also, I would suggest including `TTLStrategy.SecondsAfterCompleted` to replace `TTLSecondsAfterFinished`. This way, we can mark `TTLSecondsAfterFinished` as deprecated. To allow for backwards compatibility, we can automatically convert `TTLSecondsAfterFinished` to `TTLStrategy.SecondsAfterCompleted` during `Workflow` creation and provide a warning that it is being deprecated.
>

**Comment by NikeNano** [comment]
FYI closed this by mistake.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 7]
```diff
@@ -201,6 +208,13 @@ type WorkflowSpec struct {
 	// ttlSecondsAfterFinished expires immediately after the Workflow finishes.
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"bytes,18,opt,name=ttlSecondsAfterFinished"`
 
+	// TTLStrategy limits the lifetime of a Workflow that has finished execution depending on if it
+	// Succeeded or Failed. If this struct is set, once the Workflow finishes, it will be
+	// deleted after the time to live expires. If this field is unset,
+	// the controller config map will hold the default values
+	// Update
+	TTLStrategy TTLStrategy `json:"ttlStrategy,omitempty" protobuf:"bytes,28,opt,name=ttlStrategy"`
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: > This should be of a `*TTLStrategy` type

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 8]
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
 var _ TemplateGetter = &Workflow{}
 var _ TemplateStorage = &Workflow{}
 
+// TTLStrategy is the strategy for the time to live depending on if the workflow succeded or failed
+type TTLStrategy struct {
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: > Also, I would suggest including TTLStrategy.SecondsAfterCompleted to replace TTLSecondsAfterFinished. This way, we can mark TTLSecondsAfterFinished as deprecated. To allow for backwards compatibility, we can automatically convert TTLSecondsAfterFinished to TTLStrategy.SecondsAfterCompleted during Workflow creation and provide a warning that it is being deprecated.

**Comment by NikeNano** [comment]
@simster7 I will fix it, havn't had time to look at it during Christmas:) Where in the code base is the workflow creation? I have looked around in the operation file but not really sure where the changes are needed in order to automatically convert TTLSecondsAfterFinished to TTLStrategy.SecondsAfterCompleted during Workflow creation? 

Hope you had a great Christamas!

**Comment by NikeNano** [comment]
@simster7 I have now updated and the tests pass locally. However I have some issues understanding how to read the tests from the github action. The "default" gets me to a argo UI where I cant load the results.  The Build error gives "No configuration was found in your project. "

**Comment by alexec** [comment]
You should be able to fix build we’re by syncing with master.

**Comment by alexec** [comment]
The other error maybe a test failure. Run “make test” to check.

**Comment by NikeNano** [comment]
make test

**Comment by NikeNano** [comment]
> The other error maybe a test failure. Run “make test” to check.

Thanks! I assume you mean locally :) Running it and checking

**Comment by NikeNano** [comment]
> The other error maybe a test failure. Run “make test” to check.

Running "make test" in the main folder did not return any errors. Is there any way to get the logs from the CI run?

**Comment by NikeNano** [comment]
I have now updated according to master. Running make test locally the following test fails: argo/test/e2e it fails with No Auth Provider found for name "gcp". Is this something you are familiary with? I found this I https://github.com/kubernetes/client-go/issues/242 but don't know enough of the part of the repo to know if that makes sense .....? Is there any possibilities to get logs from the test run due to CI systems argo uses @alexec ?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 9]
```diff
@@ -201,6 +209,13 @@ type WorkflowSpec struct {
 	// ttlSecondsAfterFinished expires immediately after the Workflow finishes.
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"bytes,18,opt,name=ttlSecondsAfterFinished"`
```
[the remark this answers is not in the document]
Comment (NikeNano) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Sure I will fix this :) Was hoping to get the test running first but dont understand what the ```
hack/update-openapigen.sh
```
works, what is the goal of it? Currently get the following error 
```
API rule violation: names_match,github.com/argoproj/argo/pkg/apis/workflow/v1alpha1,TTLStrategy,Seco
```

[the remark this answers is not in the document]
Comment (simster7) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: You should make codegen using `make codegen`. You can ignore that API violation error

**Comment by NikeNano** [comment]
> Last change, otherwise LGTM!

I updated with the tag

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 10]
```diff
@@ -199,42 +207,50 @@ type WorkflowSpec struct {
...
+	// deleted after the time to live expires. If this field is unset,
+	// the controller config map will hold the default values
+	// Update
+	TTLStrategy *TTLStrategy `json:"ttlStrategy,omitempty" protobuf:"bytes,19,opt,name=ttlStrategy"`
+
 	// Optional duration in seconds relative to the workflow start time which the workflow is
 	// allowed to run before the controller terminates the workflow. A value of zero is used to
 	// terminate a Running workflow
-	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty" protobuf:"bytes,19,opt,name=activeDeadlineSeconds"`
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty" protobuf:"bytes,20,opt,name=activeDeadlineSeconds"`
```
[the remark this answers is not in the document]
Comment (NikeNano) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Ahh ok, will fix that!

[the remark this answers is not in the document]
Comment (NikeNano) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I updated now, I might be wrong but I think is should be 28, based on that this is that last entry in 
```
type WorkflowSpec struct {
 ```
 https://github.com/argoproj/argo/blob/94449876f4a571ab279802d5ca4d5e938ca3d44d/pkg/apis/workflow/v1alpha1/workflow_types.go#L237  I am new to go and protobuff so not sure though :) Also thanks for all the feedback!

**Comment by NikeNano** [comment]
> Thanks for all your work @NikeNano!

Awesome! Thanks for all the help as well @simster7!

**Pull request merged**